### DISCUSSION TITLE
Migrations are failing when PostgreSQL not created by diesel

### DISCUSSION THREAD

**Discussion by crusty-dave** [diesel-rs/diesel#3034] [Q&A] [answered]
I am using a Helm chart to create my PSQL DB in a K8s cluster.  It is using tools provided by our infrastructure team and not diesel-cli to create the DB initially and assign the correct permissions & credentials.

When I run embedded migrations or non-embedded migrations, they fail indicating that the table or column that is added by the migration already exists.

migration failed: version 20220129033212, path \"./db/config/migrations/2022-01-29-033212_tenant\", error Failed with: column \"secret_key\" of relation \"tenants\" already exists"

My work-around is to not use embedded-migrations so that I can run each migration individually using the following code:

```rust
    pub fn run_pending_migrations(conn: &PooledPgConnection, migrations_dir: &Path) {
        match diesel_migrations::mark_migrations_in_directory(conn, &migrations_dir) {
            Ok(migrations) => {
                for migration_applied in migrations {
                    let migration = &migration_applied.0;
                    let version = migration_applied.0.version();
                    let path_opt = &migration_applied.0.file_path();
                    let path = if let Some(path) = path_opt {
                        path
                    } else {
                        Path::new("<NONE>")
                    };
                    warn!(
                        "{} version {}, path {:?} applied {}",
                        PG_SWG_CONFIG_NAME, version, &path, migration_applied.1
                    );
                    if false == migration_applied.1 {
                        if let Err(e) = migration.run(conn) {
                            warn!(
                                "{} migration failed: version {}, path {:?}, error {}",
                                PG_SWG_CONFIG_NAME, version, &path, e
                            );
                        }
                    }
                }
            }
            Err(e) => {
                warn!("{} migrations error {:?}", PG_SWG_CONFIG_NAME, e);
            }
        }
    }

```

However, with the above, I still get the error, but at least I can still run a subsequent migration that needs to be run.

This is the original code using embedded-migrations that doesn't work:

```rust
        //  Make sure that our tables are up-to-date by running any
        //  pending migrations.
        warn!("{} run pending migrations", PG_SWG_CONFIG_NAME);
        if let Err(e) = embedded_migrations::run_with_output(&db, &mut std::io::stdout()) {
            let err_str = format!("Failed to run all pending migrations, error {}", e);
            warn!("{} {}", PG_SWG_CONFIG_NAME, &err_str);
        }
```

When diesel-cli creates a DB, it adds __diesel_schema_migrations that I believe is updated when the migrations are run, but I haven't found that documented anywhere on diesel.rs.  I am simply missing the documentation?  It seems like we should have the option to output that schema in the initial 00000000000000_diesel_initial_setup/up.sql that is created.

Here is what is in that table:

```sql
CREATE TABLE public.__diesel_schema_migrations
(
    version character varying(50) COLLATE pg_catalog."default" NOT NULL,
    run_on timestamp without time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
    CONSTRAINT __diesel_schema_migrations_pkey PRIMARY KEY (version)
)
```

If someone could suggest some diesel code that I could add to check to see if the table exists, and then add it if it doesn't, that would be quite helpful.

TIA,
-Dave

**Comment by weiznich** [comment, marked as answer]
The behavior described here does not match neither the implementation nor that what our tests verify. Additionally this post does not include all information required to reproduce the problem. Please provide at least the following information if your are interested in an answer:

* Used rustc, postgres and diesel version
* Used operating system
* A minimal reproducible example, including all necessary migrations.

EDIT: Generally speaking the error message indicates a misunderstanding/misuse of the way migrations work, it seems like you just apply a migration with the same content as a previous migration again. That will never work and that is considered a user error. Your "workaround" to just ignore failing migrations is considered to be bad practice for obvious reasons.

**Reply by crusty-dave** [reply]
I tried adding a migration to force the migrations table to be added, but as you suggest, it seems it already exists, so I am having another issue, at least so it would seem:

migration failed: version 20211001120000, path \"./db/config/migrations/2021-10-01-120000_migration_fix\", error Failed with: relation \"__diesel_schema_migrations\" already exists"

rustc 1.47.0 (18bf6b4f0 2020-10-07)
cargo 1.47.0 (f3c7e066a 2020-08-28)
diesel 1.4.1
Ubuntu 20.04 LTS
I think Postgres is 13.4, but I haven't been able to get psql to connect to verify.

It will take some time to put together a simplified working example.

The application is running in a docker container in a K8s cluster, so perhaps it is an environmental issue.  I don't include diesel-cli in the Docker container, so I cannot use it to check anything.  I will probably try to add that first in case it shows any dependencies that might be missed.  However, the app runs, and the migration works if I add or alter an existing table.  The issue is that the migrations always run the older migrations that should be skipped.  The documentation is quite lacking with regard to debugging issues with migration.

The issue is that while the diesel code says this migration is not applied, when it is actually applied:

version 20211001223956, path \"./db/config/migrations/2021-10-01-223956_create\" applied false"

Hence the error.

Is there something specific that happens when a DB is created with diesel? 

When I run diesel-cli in a WSL Ubuntu 20.04 LTS running on Windows with Postgres 13.4, I don't see the issue.  So I believe that it may be something with the container environment.

Here are the packages that are being included:

FROM ubuntu:20.04
RUN apt update \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends postgresql-client \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends libpq-dev \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends openssl \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends libssl-dev \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends mlocate \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends net-tools \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends iputils-ping \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends procinfo \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends pkg-config \
    && true

Thanks,
-Dave

**Reply by crusty-dave** [reply]
FYI: I am not running migrations from build.rs, I would assume that is not required, but it is referenced in the documentation:

"As an alternative to running migrations with the CLI, you can call diesel::migrations::run_pending_migrations from build.rs."

**Reply by weiznich** [reply]
All three variants (CLI, `diesel_mgirations::run_pending_migrations` and  `embed_migrations!`) to run migrations call internally the same code, so there is no reason that they produce a different result, at least not without calling them with entirely different arguments.

>  The issue is that the migrations always run the older migrations that should be skipped. 

Specifically that's not the case. Older migrations are not executed as long as you executed them before with any of the methods mentioned above. If that is not the case for you that could have the following reasons:

* You've executed the SQL manually, so that our migration system cannot know about that
* You've made manual modifications to your database schema or the `__diesel_schema_migrations` table
* You've renamed a migration

> The documentation is quite lacking with regard to debugging issues with migration.

There is just not much to document here: If a migration is applied the corresponding version appears in `__diesel_schema_migrations` . Any future call to one of the `run_*_migrations` functions will skip all migrations with versions in `__diesel_schema_migrations`. The only exception here is if you manually call undocumented low level functions in the migrations module of diesel or in `diesel_migrations`, but in that case you are just in territory where we don't guarantee anything and don't provide any support. If that seems not to happen for you that likely indicates some issue with your setup, which again is something we cannot help you without reproducible example. So again: Please provide a reproducible example, otherwise I cannot help you any further. That's likely my last answer here, til the relevant information is provided.

**Reply by crusty-dave** [reply]
I reduced my r2d2 connection pool, then I was able to connect using psql to the DB, the version is much older than I thought, could I be dealing with an issue with an older version of PostgreSQL?

psql (12.9 (Ubuntu 12.9-0ubuntu0.20.04.1), server 10.5 (Ubuntu 10.5-1.pgdg18.04+1))
SSL connection (protocol: TLSv1.2, cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256, compression: off)

I hadn't touched my schema since October until this last week, and I had been seeing those migration errors all along.  I did delete a migration way back in September when I moved to the October migration.  Are the migration versions built into the code, or just used after the DB is created?  The DB that I am currently using was created in January.

I really want to be able to switch back to the embedded_migrations, so I will see if I can come up with an example using that.

Thanks,
-Dave

**Reply by crusty-dave** [reply]
I tried running on two other VMs and I didn't see the issue.

That caused me to review my steps, apparently, when I had a bad build, I ran dbdrop=true, but it was with the same bad build, so it recreated the issue.  Once I ran dbdrop=true with a good build, it fixed the issue and the embedded_migrartions are now working correctly.

Sorry for the SPAM and thanks for your patience.

My apologies,
-Dave

**Reply by weiznich** [reply]
Thanks for coming back and  writing a post that the issue is now resolved. Such things happens for all of us at some point, so no worries about that. Generally speaking: Such cases are the reason I tend to ask for reproducible examples as it is otherwise impossible to understand what's going on.
